Limber/MIXMockServer.cpp

```diff
@@ -0,0 +1,274 @@
+
+using namespace Swift;
+
+class SimpleMIXChannelRegistry {
+
+    public:
+        SimpleMIXChannelRegistry() {}
+
+        void addMIXChannel(const JID& channelJID) {
+            channels_.insert(channelJID);
+        }
+
+        const std::unordered_set<std::string>& getChannels() {
+            return channels_;
+        }
+
+        bool hasMIXChannel(const JID& jid) {
+            return channels_.find(jid) != channels_.end();
+        }
+
+    private:
+        std::unordered_set<std::string> channels_;
+};
+
+class Server {
+    public:
+        Server(UserRegistry* userRegistry, SimpleMIXChannelRegistry* mixChannelRegistry, EventLoop* eventLoop) : userRegistry_(userRegistry), mixChannelRegistry_(mixChannelRegistry) {
+            serverFromClientConnectionServer_ = BoostConnectionServer::create(5222, boostIOServiceThread_.getIOService(), eventLoop);
+            serverFromClientConnectionServer_->onNewConnection.connect(boost::bind(&Server::handleNewConnection, this, _1));
+            serverFromClientConnectionServer_->start();
+        }
+
+    private:
+        void handleNewConnection(std::shared_ptr<Connection> c) {
+            auto session  = std::make_shared<ServerFromClientSession>(idGenerator_.generateID(), c, &payloadParserFactories_, &payloadSerializers_, &xmlParserFactory_, userRegistry_);
+            serverFromClientSessions_.push_back(session);
+            session->onElementReceived.connect(boost::bind(&Server::handleElementReceived, this, _1, session));
+            session->onSessionFinished.connect(boost::bind(&Server::handleSessionFinished, this, session));
+            session->startSession();
+        }
+
+        void handleSessionFinished(std::shared_ptr<ServerFromClientSession> session) {
+            serverFromClientSessions_.erase(std::remove(serverFromClientSessions_.begin(), serverFromClientSessions_.end(), session), serverFromClientSessions_.end());
+        }
+
+        void handleElementReceived(std::shared_ptr<ToplevelElement> element, std::shared_ptr<ServerFromClientSession> session) {
+            std::shared_ptr<Stanza> stanza(std::dynamic_pointer_cast<Stanza>(element));
+            if (!stanza) {
+                return;
+            }
+            stanza->setFrom(session->getRemoteJID());
+
+            auto iq = std::dynamic_pointer_cast<IQ>(stanza);
+            if (!iq) {
+                return;
+            }
+
+            if (mixChannelRegistry_->hasMIXChannel(stanza->getTo())) {
+                //If request comes to particular channel supported by service.
+                handleElementReceivedForMIXChannel(iq, session);
+            } else if (stanza->getTo() == session->getLocalJID()) {
+                //If request comes to domain service
+                handleElementReceivedForMIXService(iq, session);
+            } else if (stanza->getTo() == session->getRemoteJID().toBare() || stanza->getTo() == JID()) {
+                //If request comes to own local server
```

Comment:
This is to the account and not to the local server, the local server would be `to='server domain'`.